x/stake/keeper.go

```diff
@@ -33,319 +29,356 @@ func NewKeeper(cdc *wire.Codec, key sdk.StoreKey, ck bank.Keeper, codespace sdk.
+	b := store.Get(GetValidatorKey(addr))
 	if b == nil {
-		return 0
-	}
-	var counter int16
-	err := k.cdc.UnmarshalBinary(b, &counter)
-	if err != nil {
-		panic(err)
+		return validator, false
 	}
-	return counter
+	k.cdc.MustUnmarshalBinary(b, &validator)
+	return validator, true
 }
 
-// set the current in-block validator operation counter
-func (k Keeper) setCounter(ctx sdk.Context, counter int16) {
+// Get the set of all validators with no limits, used during genesis dump
+func (k Keeper) getAllValidators(ctx sdk.Context) (validators Validators) {
 	store := ctx.KVStore(k.storeKey)
-	bz, err := k.cdc.MarshalBinary(counter)
-	if err != nil {
-		panic(err)
-	}
-	store.Set(CounterKey, bz)
-}
+	iterator := store.SubspaceIterator(ValidatorsKey)
 
-//_________________________________________________________________________
-
-// get a single candidate
-func (k Keeper) GetCandidate(ctx sdk.Context, addr sdk.Address) (candidate Candidate, found bool) {
-	store := ctx.KVStore(k.storeKey)
-	b := store.Get(GetCandidateKey(addr))
-	if b == nil {
-		return candidate, false
-	}
-	err := k.cdc.UnmarshalBinary(b, &candidate)
-	if err != nil {
-		panic(err)
+	i := 0
+	for ; ; i++ {
+		if !iterator.Valid() {
+			iterator.Close()
+			break
+		}
+		bz := iterator.Value()
+		var validator Validator
+		k.cdc.MustUnmarshalBinary(bz, &validator)
+		validators = append(validators, validator)
+		iterator.Next()
 	}
-	return candidate, true
+	return validators[:i] // trim
 }
 
-// Get the set of all candidates, retrieve a maxRetrieve number of records
-func (k Keeper) GetCandidates(ctx sdk.Context, maxRetrieve int16) (candidates Candidates) {
+// Get the set of all validators, retrieve a maxRetrieve number of records
+func (k Keeper) GetValidators(ctx sdk.Context, maxRetrieve int16) (validators Validators) {
 	store := ctx.KVStore(k.storeKey)
-	iterator := store.Iterator(subspace(CandidatesKey))
+	iterator := store.SubspaceIterator(ValidatorsKey)
 
-	candidates = make([]Candidate, maxRetrieve)
+	validators = make([]Validator, maxRetrieve)
 	i := 0
 	for ; ; i++ {
 		if !iterator.Valid() || i > int(maxRetrieve-1) {
 			iterator.Close()
 			break
 		}
 		bz := iterator.Value()
-		var candidate Candidate
-		err := k.cdc.UnmarshalBinary(bz, &candidate)
-		if err != nil {
-			panic(err)
-		}
-		candidates[i] = candidate
+		var validator Validator
+		k.cdc.MustUnmarshalBinary(bz, &validator)
+		validators[i] = validator
 		iterator.Next()
 	}
-	return candidates[:i] // trim
+	return validators[:i] // trim
 }
 
-func (k Keeper) setCandidate(ctx sdk.Context, candidate Candidate) {
+func (k Keeper) setValidator(ctx sdk.Context, validator Validator) Validator {
 	store := ctx.KVStore(k.storeKey)
-	address := candidate.Address
+	pool := k.getPool(store)
+	address := validator.Address
 
-	// retreive the old candidate record
-	oldCandidate, oldFound := k.GetCandidate(ctx, address)
+	// update the main list ordered by address before exiting
+	defer func() {
```

Comment:
Why use defer here?

Reply:
There are three endpoints in this function, they all need to write to this store as the final step. The alternative is to just expand and implement 3 times. I like the defer, but don't feel very strongly about it